source/common/router/router.cc

```diff
@@ -1381,6 +1417,23 @@ Filter::UpstreamRequest::UpstreamRequest(Filter& parent, Http::ConnectionPool::I
     }
   }
 
+  // If configured, set up the per-try histogram to fire when the UpstreamRequest
+  // completes. This is done via a Cleanup so that it can be cancelled if the global
+  // timeout is what cancels the request instead of the per-try timeout, and then avoid
+  // writing a misleading value into the histogram.
```

Comment:
I get why you are doing this, but are there really more cases other than per-try timeout and upstream complete? Can't we just call a utility function directly in those cases, with the per-try timeout case passing 100 per my comment above? It seems like if we don't need a Cleanup in the global case we don't need one here. The reason I keep coming back to this is this code is already incredibly complicated and magical stuff like this makes it even harder to grok the control flow.

Reply:
I think it would be cleaner for the global to use a Cleanup as well since it has two different places to fire mostly identical code.  It doesn't need the cancellable aspect of the Cleanup, so I put it in two places to avoid leaning on RAII magic.

I don't have better ideas for how to keep the per-try firing cancellable than the Cleanup's method of assigning to a function that's called during the per-try destructor.  Is it less magic to set a member std::function variable to `[](){}` rather than call a named member function named cancel?

Reply:
Slept on it. I’ll change this today.

Reply:
Great, thanks. 

/wait